File: results/0.25/MQTT/mqtt-claude-3-5-haiku-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdbool.h>

typedef enum {
    CONNECT = 1,
    CONNACK = 2,
    PUBLISH = 3,
    PUBACK = 4,
    PUBREC = 5,
    PUBREL = 6,
    PUBCOMP = 7,
    SUBSCRIBE = 8,
    SUBACK = 9,
    UNSUBSCRIBE = 10,
    UNSUBACK = 11,
    PINGREQ = 12,
    PINGRESP = 13,
    DISCONNECT = 14
} MQTTPacketType;

typedef struct {
    MQTTPacketType type;
    uint8_t flags;
    HParser* parser;
} MQTTPacket;

HParser* mqtt_protocol_name() {
    return h_sequence(
        h_ch('M'), h_ch('Q'), h_ch('T'), h_ch('T'),
        NULL
    );
}

HParser* mqtt_variable_length_integer() {
    return h_many1(h_uint8());
}

HParser* mqtt_utf8_string() {
    HParser* length = h_uint16();
    HParser* string = h_length_value(length, h_ch_range(0x20, 0x7E));
    return string;
}

HParser* mqtt_properties() {
    return h_optional(
        h_sequence(
            h_uint8(),  // Property length
            h_many(
                h_choice(
                    h_uint8(),  // Property identifier
                    h_choice(
                        mqtt_utf8_string(),
                        h_uint32(),
                        h_uint16(),
                        NULL
                    ),
                    NULL
                )
            ),
            NULL
        )
    );
}

HParser* mqtt_connect_parser() {
    return h_sequence(
        mqtt_protocol_name(),
        h_uint8(),  // Protocol version
        h_bits(8, false),  // Connect flags
        h_uint16(),  // Keep alive
        mqtt_properties(),
        mqtt_utf8_string(),  // Client ID
        h_optional(mqtt_utf8_string()),  // Username
        h_optional(mqtt_utf8_string()),  // Password
        NULL
    );
}

HParser* mqtt_publish_parser() {
    return h_sequence(
        mqtt_utf8_string(),  // Topic name
        h_optional(h_uint16()),  // Packet identifier
        mqtt_properties(),
        h_many(h_uint8()),  // Payload
        NULL
    );
}

HParser* mqtt_packet_parser() {
    return h_choice(
        mqtt_connect_parser(),
        mqtt_publish_parser(),
        NULL
    );
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <mqtt_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(buffer, 1, file_size, file);
    fclose(file);

    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        return 1;
    }

    HParser* parser = mqtt_packet_parser();
    HParseResult* result = h_parse(parser, buffer, read_size);

    if (result && result->ast) {
        printf("MQTT packet parsed successfully\n");
    } else {
        printf("MQTT packet parsing failed\n");
    }

    h_parse_result_free(result);
    h_parser_free(parser);
    free(buffer);

    return 0;
}

File: results/0.25/MQTT/mqtt-claude-3-5-haiku-20241022.rs
use nom::{
    bytes::complete::{tag, take},
    combinator::{map, opt},
    multi::{count, many0},
    number::complete::{be_u8, be_u16, be_u32},
    sequence::{tuple, preceded},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
enum MQTTPacketType {
    Connect,
    ConnAck,
    Publish,
    PubAck,
    PubRec,
    PubRel,
    PubComp,
    Subscribe,
    SubAck,
    Unsubscribe,
    UnsubAck,
    PingReq,
    PingResp,
    Disconnect,
}

#[derive(Debug)]
struct MQTTPacket {
    packet_type: MQTTPacketType,
    flags: u8,
    remaining_length: usize,
    payload: Option<Vec<u8>>,
}

fn parse_mqtt_packet(input: &[u8]) -> IResult<&[u8], MQTTPacket> {
    let (input, first_byte) = be_u8(input)?;
    let packet_type = match first_byte >> 4 {
        1 => MQTTPacketType::Connect,
        2 => MQTTPacketType::ConnAck,
        3 => MQTTPacketType::Publish,
        4 => MQTTPacketType::PubAck,
        5 => MQTTPacketType::PubRec,
        6 => MQTTPacketType::PubRel,
        7 => MQTTPacketType::PubComp,
        8 => MQTTPacketType::Subscribe,
        9 => MQTTPacketType::SubAck,
        10 => MQTTPacketType::Unsubscribe,
        11 => MQTTPacketType::UnsubAck,
        12 => MQTTPacketType::PingReq,
        13 => MQTTPacketType::PingResp,
        14 => MQTTPacketType::Disconnect,
        _ => return Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::Tag))),
    };
    let flags = first_byte & 0x0F;

    let (input, remaining_length) = parse_remaining_length(input)?;

    let (input, payload) = if remaining_length > 0 {
        let (input, payload_bytes) = take(remaining_length)(input)?;
        (input, Some(payload_bytes.to_vec()))
    } else {
        (input, None)
    };

    Ok((input, MQTTPacket {
        packet_type,
        flags,
        remaining_length,
        payload,
    }))
}

fn parse_remaining_length(input: &[u8]) -> IResult<&[u8], usize> {
    let mut multiplier = 1;
    let mut value = 0;
    let mut bytes_read = 0;

    loop {
        let (new_input, encoded_byte) = be_u8(input)?;
        input = new_input;
        value += (encoded_byte & 0x7F) as usize * multiplier;
        multiplier *= 128;
        bytes_read += 1;

        if encoded_byte & 0x80 == 0 {
            break;
        }

        if bytes_read > 4 {
            return Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::TooLarge)));
        }
    }

    Ok((input, value))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <mqtt_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_mqtt_packet(&buffer) {
        Ok((_, packet)) => {
            println!("Parsed MQTT Packet: {:?}", packet);
        }
        Err(e) => {
            eprintln!("Failed to parse MQTT packet: {:?}", e);
            std::process::exit(1);
        }
    }

    Ok(())
}